pkg/ebpf/c/tracee.bpf.c

```diff
@@ -5153,73 +5182,175 @@ CGROUP_SKB_HANDLE_FUNCTION(proto_icmpv6);
             size = ctx->len;
             break;
-        default:
-            size += neteventctx->md.header_size;       // add headers size
-            size = size > ctx->len ? ctx->len : size;  // check limits
+        default: // submit size bytes
+            size += neteventctx->md.header_size;
+            size = size > ctx->len ? ctx->len : size;
             break;
     }
 
-    flags |= (u64) size << 32;
+    // Flag eBPF subsystem to use current CPU and copy size bytes of payload.
+    u64 flags = BPF_F_CURRENT_CPU | (u64) size << 32;
     neteventctx->bytes = size;
 
-    // set the event type before submitting event
+    // Set the event type before submitting event.
     neteventctx->eventctx.eventid = event_type;
 
-    return bpf_perf_event_output(ctx,
-                                 map,
-                                 flags,
-                                 neteventctx,
-                                 sizeof_net_event_context_t());
+    // Submit the event.
+    return bpf_perf_event_output(ctx, map, flags, neteventctx, sizeof_net_event_context_t());
 }
 
-#define cgroup_skb_submit_event(a,b,c,d) cgroup_skb_submit(&events,a,b,c,d)
+// Submit a network event.
+#define cgroup_skb_submit_event(a, b, c, d) cgroup_skb_submit(&events, a, b, c, d)
+
+// Check if a flag is set in the retval.
+#define retval_hasflag(flag) (neteventctx->eventctx.retval & flag) == flag
+
+// Keep track of a flow event if they are enabled and if any policy matched.
+// Submit the flow base event so userland can derive the flow events.
+statfunc u32 cgroup_skb_submit_flow(struct __sk_buff *ctx,
+                                    net_event_context_t *neteventctx,
+                                    u32 event_type, u32 size, u32 flow)
+{
+    netflowvalue_t *netflowvalptr, netflowvalue = {
+                                       .last_update = bpf_ktime_get_ns(),
+                                       .direction = flow_unknown,
+                                   };
+
+    // Set the current netctx task as the flow task.
+    neteventctx->md.flow.host_pid = neteventctx->eventctx.task.host_pid;
+
+    // Set the flow event type in retval.
+    neteventctx->eventctx.retval |= flow;
+
+    // Check if the current packet source is the flow initiator.
+    bool is_initiator = 0;
+
+    switch (flow) {
+        // 1) TCP connection is being established.
+        case flow_tcp_begin:
+            // Ingress: Remote (src) is sending SYN+ACK: this host (dst) is the initiator.
+            if (retval_hasflag(packet_ingress))
+                netflowvalue.direction = flow_outgoing;
+
+            // Egress: Host (src) is sending SYN+ACK: remote (dst) host is the initiator.
+            if (retval_hasflag(packet_egress))
+                netflowvalue.direction = flow_incoming;
+
+            // Invert src/dst: The flowmap src should always be set to flow initiator.
+            neteventctx->md.flow = invert_netflow(neteventctx->md.flow);
```

Comment:
Should this be in both cases? This seems wrong if the flow is outgoing (this host is the already the src, so why invert it?)

Reply:
I believe this was answered in private when we spoke about SYN -> SYN+ACK -> ACK (and that there would be only a single SYN+ACK, and a flag telling me if the src of the packet initiated the flow or not).